doc/bash.1
.\"
.\" MAN PAGE COMMENTS to
.\"
.\"	Chet Ramey
.\"	Information Network Services
.\"	Case Western Reserve University
.\"	chet@po.cwru.edu
.\"
.\"	Last Change: Thu Sep 28 10:25:59 EDT 2006
.\"
.\" bash_builtins, strip all but Built-Ins section
.if \n(zZ=1 .ig zZ
.if \n(zY=1 .ig zY
.TH BASH 1 "2006 September 28" "GNU Bash-3.2"
.\"
.\" There's some problem with having a `@'
.\" in a tagged paragraph with the BSD man macros.
.\" It has to do with `@' appearing in the }1 macro.
.\" This is a problem on 4.3 BSD and Ultrix, but Sun
.\" appears to have fixed it.
.\" If you're seeing the characters
.\" `@u-3p' appearing before the lines reading
.\" `possible-hostname-completions
.\" and `complete-hostname' down in READLINE,
.\" then uncomment this redefinition.
.\"
.de }1
.ds ]X \&\\*(]B\\
.nr )E 0
.if !"\\$1"" .nr )I \\$1n
.}f
.ll \\n(LLu
.in \\n()Ru+\\n(INu+\\n()Iu
.ti \\n(INu
.ie !\\n()Iu+\\n()Ru-\w\\*(]Xu-3p \{\\*(]X
.br\}
.el \\*(]X\h|\\n()Iu+\\n()Ru\c
.}f
..
.\"
.\" File Name macro.  This used to be `.PN', for Path Name,
.\" but Sun doesn't seem to like that very much.
.\"
.de FN
\fI\|\\$1\|\fP
..
.SH NAME
bash \- GNU Bourne-Again SHell
.SH SYNOPSIS
.B bash
[options]
[file]
.SH COPYRIGHT
.if n Bash is Copyright (C) 1989-2005 by the Free Software Foundation, Inc.
.if t Bash is Copyright \(co 1989-2005 by the Free Software Foundation, Inc.
.SH DESCRIPTION
.B Bash
is an \fBsh\fR-compatible command language interpreter that
executes commands read from the standard input or from a file.
.B Bash
also incorporates useful features from the \fIKorn\fP and \fIC\fP
shells (\fBksh\fP and \fBcsh\fP).
.PP
.B Bash
is intended to be a conformant implementation of the
Shell and Utilities portion of the IEEE POSIX specification
(IEEE Standard 1003.1).
.B Bash
can be configured to be POSIX-conformant by default.
.SH OPTIONS
In addition to the single-character shell options documented in the
description of the \fBset\fR builtin command, \fBbash\fR
interprets the following options when it is invoked:
.PP
.PD 0
.TP 10
.BI \-c "\| string\^"
If the
.B \-c
option is present, then commands are read from
.IR string .
If there are arguments after the
.IR string ,
they are assigned to the positional parameters, starting with
.BR $0 .
.TP
.B \-i
If the
.B \-i
option is present, the shell is
.IR interactive .
.TP
.B \-l
Make
.B bash
act as if it had been invoked as a login shell (see
.SM
.B INVOCATION
below).
.TP
.B \-r
If the
.B \-r
option is present, the shell becomes
.I restricted
(see
.SM
.B "RESTRICTED SHELL"
below).
.TP
.B \-s
If the
.B \-s
option is present, or if no arguments remain after option
processing, then commands are read from the standard input.
This option allows the positional parameters to be set
when invoking an interactive shell.
.TP
.B \-D
A list of all double-quoted strings preceded by \fB$\fP
is printed on the standard output.
These are the strings that
are subject to language translation when the current locale
is not \fBC\fP or \fBPOSIX\fP.
This implies the \fB\-n\fP option; no commands will be executed.
.TP
.B [\-+]O [\fIshopt_option\fP]
\fIshopt_option\fP is one of the shell options accepted by the
\fBshopt\fP builtin (see
.SM
.B SHELL BUILTIN COMMANDS
below).
If \fIshopt_option\fP is present, \fB\-O\fP sets the value of that option;
\fB+O\fP unsets it.
If \fIshopt_option\fP is not supplied, the names and values of the shell
options accepted by \fBshopt\fP are printed on the standard output.
If the invocation option is \fB+O\fP, the output is displayed in a format
that may be reused as input.
.TP
.B \-\-
A
.B \-\-
signals the end of options and disables further option processing.
Any arguments after the
.B \-\-
are treated as filenames and arguments.  An argument of
.B \-
is equivalent to \fB\-\-\fP.
.PD
.PP
.B Bash
also interprets a number of multi-character options.
These options must appear on the command line before the
single-character options to be recognized.
.PP
.PD 0
.TP
.B \-\-debugger
Arrange for the debugger profile to be executed before the shell
starts.
Turns on extended debugging mode (see the description of the
.B extdebug
option to the
.B shopt
builtin below)
and shell function tracing (see the description of the
\fB\-o functrace\fP option to the
.B set
builtin below).
.TP
.B \-\-dump\-po\-strings
Equivalent to \fB\-D\fP, but the output is in the GNU \fIgettext\fP
\fBpo\fP (portable object) file format.
.TP
.B \-\-dump\-strings
Equivalent to \fB\-D\fP.
.TP
.B \-\-help
Display a usage message on standard output and exit successfully.
.TP
\fB\-\-init\-file\fP \fIfile\fP
.PD 0
.TP
\fB\-\-rcfile\fP \fIfile\fP
.PD
Execute commands from
.I file
instead of the system wide initialization file
.I /etc/bash.bashrc
and the standard personal initialization file
.I ~/.bashrc
if the shell is interactive (see
.SM
.B INVOCATION
below).
.TP
.B \-\-login
Equivalent to \fB\-l\fP.
.TP
.B \-\-noediting
Do not use the GNU
.B readline
library to read command lines when the shell is interactive.
.TP
.B \-\-noprofile
Do not read either the system-wide startup file
.FN /etc/profile
or any of the personal initialization files
.IR ~/.bash_profile ,
.IR ~/.bash_login ,
or
.IR ~/.profile .
By default,
.B bash
reads these files when it is invoked as a login shell (see
.SM
.B INVOCATION
below).
.TP
.B \-\-norc
Do not read and execute the system wide initialization file
.I /etc/bash.bashrc
and the personal initialization file
.I ~/.bashrc
if the shell is interactive.
This option is on by default if the shell is invoked as
.BR sh .
.TP
.B \-\-posix
Change the behavior of \fBbash\fP where the default operation differs
from the POSIX standard to match the standard (\fIposix mode\fP).
.TP
.B \-\-restricted
The shell becomes restricted (see
.SM
.B "RESTRICTED SHELL"
below).
.TP
.B \-\-verbose
Equivalent to  \fB\-v\fP.
.TP
.B \-\-version
Show version information for this instance of
.B bash
on the standard output and exit successfully.
.PD
.SH ARGUMENTS
If arguments remain after option processing, and neither the
.B \-c
nor the
.B \-s
option has been supplied, the first argument is assumed to
be the name of a file containing shell commands.
If
.B bash
is invoked in this fashion, 
.B $0
is set to the name of the file, and the positional parameters
are set to the remaining arguments.
.B Bash
reads and executes commands from this file, then exits.
\fBBash\fP's exit status is the exit status of the last command
executed in the script.
If no commands are executed, the exit status is 0.
An attempt is first made to open the file in the current directory, and,
if no file is found, then the shell searches the directories in
.SM
.B PATH
for the script.
.SH INVOCATION
A \fIlogin shell\fP is one whose first character of argument zero is a
.BR \- ,
or one started with the 
.B \-\-login
option.
.PP
An \fIinteractive\fP shell is one started without non-option arguments
and without the
.B \-c
option
whose standard input and error are
both connected to terminals (as determined by
.IR isatty (3)),
or one started with the
.B \-i
option.
.SM
.B PS1
is set and
.B $\-
includes
.B i
if
.B bash
is interactive,
allowing a shell script or a startup file to test this state.
.PP
The following paragraphs describe how
.B bash
executes its startup files.
If any of the files exist but cannot be read,
.B bash
reports an error.
Tildes are expanded in file names as described below under
.B "Tilde Expansion"
in the
.SM
.B EXPANSION
section.
.PP
When
.B bash
is invoked as an interactive login shell, or as a non-interactive shell
with the \fB\-\-login\fP option, it first reads and
executes commands from the file \fI/etc/profile\fP, if that
file exists.
After reading that file, it looks for \fI~/.bash_profile\fP,
\fI~/.bash_login\fP, and \fI~/.profile\fP, in that order, and reads
and executes commands from the first one that exists and is readable.
The
.B \-\-noprofile
option may be used when the shell is started to inhibit this behavior.
.PP
When a login shell exits,
.B bash
reads and executes commands from the file \fI~/.bash_logout\fP, if it
exists.
.PP
When an interactive shell that is not a login shell is started,
.B bash
reads and executes commands from \fI/etc/bash.bashrc\fP and \fI~/.bashrc\fP,
if these files exist.
This may be inhibited by using the
.B \-\-norc
option.
The \fB\-\-rcfile\fP \fIfile\fP option will force
.B bash
to read and execute commands from \fIfile\fP instead of
\fI/etc/bash.bashrc\fP and \fI~/.bashrc\fP.
.PP
When
.B bash
is started non-interactively, to run a shell script, for example, it
looks for the variable
.SM
.B BASH_ENV
in the environment, expands its value if it appears there, and uses the
expanded value as the name of a file to read and execute.
.B Bash
behaves as if the following command were executed:
.sp .5
.RS
.if t \f(CWif [ \-n "$BASH_ENV" ]; then . "$BASH_ENV"; fi\fP
.if n if [ \-n "$BASH_ENV" ]; then . "$BASH_ENV"; fi
.RE
.sp .5
but the value of the
.SM
.B PATH
variable is not used to search for the file name.
.PP
If
.B bash
is invoked with the name
.BR sh ,
it tries to mimic the startup behavior of historical versions of
.B sh
as closely as possible,
while conforming to the POSIX standard as well.
When invoked as an interactive login shell, or a non-interactive
shell with the \fB\-\-login\fP option, it first attempts to
read and execute commands from
.I /etc/profile
and
.IR ~/.profile ,
in that order.
The
.B \-\-noprofile
option may be used to inhibit this behavior.
When invoked as an interactive shell with the name
.BR sh ,
.B bash
looks for the variable
.SM
.BR ENV ,
expands its value if it is defined, and uses the
expanded value as the name of a file to read and execute.
Since a shell invoked as
.B sh
does not attempt to read and execute commands from any other startup
files, the
.B \-\-rcfile
option has no effect.
A non-interactive shell invoked with the name
.B sh
does not attempt to read any other startup files. 
When invoked as
.BR sh ,
.B bash
enters
.I posix
mode after the startup files are read.
.PP
When
.B bash
is started in
.I posix
mode, as with the
.B \-\-posix
command line option, it follows the POSIX standard for startup files.
In this mode, interactive shells expand the
.SM
.B ENV
variable and commands are read and executed from the file
whose name is the expanded value.
No other startup files are read.
.PP
.B Bash
attempts to determine when it is being run by the remote shell
daemon, usually \fIrshd\fP.
If
.B bash
determines it is being run by \fIrshd\fP, it reads and executes
commands from \fI/etc/bash.bashrc\fP and \fI~/.bashrc\fP,
if these files exist and are readable.
It will not do this if invoked as \fBsh\fP.
The
.B \-\-norc
option may be used to inhibit this behavior, and the
.B \-\-rcfile
option may be used to force another file to be read, but
\fIrshd\fP does not generally invoke the shell with those options
or allow them to be specified.
.PP
If the shell is started with the effective user (group) id not equal to the
real user (group) id, and the \fB\-p\fP option is not supplied, no startup
files are read, shell functions are not inherited from the environment, the
.SM
.B SHELLOPTS
variable, if it appears in the environment, is ignored,
and the effective user id is set to the real user id.
If the \fB\-p\fP option is supplied at invocation, the startup behavior is
the same, but the effective user id is not reset.
.SH DEFINITIONS
.PP
The following definitions are used throughout the rest of this
document.
.PD 0
.TP
.B blank 
A space or tab.
.TP
.B word
A sequence of characters considered as a single unit by the shell.
Also known as a
.BR token .
.TP
.B name
A 
.I word
consisting only of alphanumeric characters and underscores, and
beginning with an alphabetic character or an underscore.  Also
referred to as an
.BR identifier .
.TP
.B metacharacter
A character that, when unquoted, separates words.  One of the following:
.br
.RS
.PP
.if t \fB|  &  ;  (  )  <  >  space  tab\fP
.if n \fB|  & ; ( ) < > space tab\fP
.RE
.PP
.TP
.B control operator
A \fItoken\fP that performs a control function.  It is one of the following
symbols:
.RS
.PP
.if t \fB\(bv\(bv  &  &&  ;  ;;  (  )  |  <newline>\fP
.if n \fB|| & && ; ;; ( ) | <newline>\fP
.RE
.PD
.SH "RESERVED WORDS"
\fIReserved words\fP are words that have a special meaning to the shell.
The following words are recognized as reserved when unquoted and either
the first word of a simple command (see
.SM
.B SHELL GRAMMAR
below) or the third word of a 
.B case 
or
.B for
command:
.if t .RS
.PP
.B
.if n ! case  do done elif else esac fi for function if in select then until while { } time [[ ]]
.if t !    case    do    done    elif    else    esac    fi    for    function    if    in    select    then    until    while    {    }    time    [[    ]]
.if t .RE
.SH "SHELL GRAMMAR"
.SS Simple Commands
.PP
A \fIsimple command\fP is a sequence of optional variable assignments
followed by \fBblank\fP-separated words and redirections, and
terminated by a \fIcontrol operator\fP.  The first word
specifies the command to be executed, and is passed as argument zero.
The remaining words are passed as arguments to the invoked command.
.PP
The return value of a \fIsimple command\fP is its exit status, or
128+\fIn\^\fP if the command is terminated by signal
.IR n .
.SS Pipelines
.PP
A \fIpipeline\fP is a sequence of one or more commands separated by
the character
.BR | .
The format for a pipeline is:
.RS
.PP
[\fBtime\fP [\fB\-p\fP]] [ ! ] \fIcommand\fP [ \fB|\fP \fIcommand2\fP ... ]
.RE
.PP
The standard output of
.I command
is connected via a pipe to the standard input of
.IR command2 .
This connection is performed before any redirections specified by the
command (see
.SM
.B REDIRECTION
below).
.PP
The return status of a pipeline is the exit status of the last
command, unless the \fBpipefail\fP option is enabled.
If \fBpipefail\fP is enabled, the pipeline's return status is the
value of the last (rightmost) command to exit with a non-zero status,
or zero if all commands exit successfully.
If the reserved word
.B !
precedes a pipeline, the exit status of that pipeline is the logical
negation of the exit status as described above.
The shell waits for all commands in the pipeline to
terminate before returning a value.
.PP
If the
.B time
reserved word precedes a pipeline, the elapsed as well as user and
system time consumed by its execution are reported when the pipeline
terminates.
The \fB\-p\fP option changes the output format to that specified by POSIX.
The
.SM
.B TIMEFORMAT
variable may be set to a format string that specifies how the timing
information should be displayed; see the description of
.SM
.B TIMEFORMAT
under
.B "Shell Variables"
below.
.PP
Each command in a pipeline is executed as a separate process (i.e., in a
subshell).
.SS Lists
.PP
A \fIlist\fP is a sequence of one or more pipelines separated by one
of the operators
.BR ; ,
.BR & ,
.BR && ,
or
.BR \(bv\(bv ,
and optionally terminated by one of
.BR ; ,
.BR & ,
or
.BR <newline> .
.PP
Of these list operators,
.B &&
and
.B \(bv\(bv
have equal precedence, followed by
.B ;
and
.BR &,
which have equal precedence.
.PP
A sequence of one or more newlines may appear in a \fIlist\fP instead
of a semicolon to delimit commands.
.PP
If a command is terminated by the control operator
.BR & ,
the shell executes the command in the \fIbackground\fP
in a subshell.  The shell does not wait for the command to
finish, and the return status is 0.  Commands separated by a
.B ;
are executed sequentially; the shell waits for each
command to terminate in turn.  The return status is the
exit status of the last command executed.
.PP
The control operators
.B &&
and
.B \(bv\(bv
denote AND lists and OR lists, respectively.
An AND list has the form
.RS
.PP
\fIcommand1\fP \fB&&\fP \fIcommand2\fP
.RE
.PP
.I command2
is executed if, and only if,
.I command1
returns an exit status of zero.
.PP
An OR list has the form
.RS
.PP
\fIcommand1\fP \fB\(bv\(bv\fP \fIcommand2\fP
.PP
.RE
.PP
.I command2
is executed if and only if
.I command1
returns a non-zero exit status.  The return status of
AND and OR lists is the exit status of the last command
executed in the list.
.SS Compound Commands
.PP
A \fIcompound command\fP is one of the following:
.TP
(\fIlist\fP)
\fIlist\fP is executed in a subshell environment (see
.SM
\fBCOMMAND EXECUTION ENVIRONMENT\fP
below).
Variable assignments and builtin
commands that affect the shell's environment do not remain in effect
after the command completes.  The return status is the exit status of
\fIlist\fP.
.TP
{ \fIlist\fP; }
\fIlist\fP is simply executed in the current shell environment.
\fIlist\fP must be terminated with a newline or semicolon.
This is known as a \fIgroup command\fP.
The return status is the exit status of
\fIlist\fP.
Note that unlike the metacharacters \fB(\fP and \fB)\fP, \fB{\fP and
\fB}\fP are \fIreserved words\fP and must occur where a reserved
word is permitted to be recognized.  Since they do not cause a word
break, they must be separated from \fIlist\fP by whitespace.
.TP
((\fIexpression\fP))
The \fIexpression\fP is evaluated according to the rules described
below under
.SM
.BR "ARITHMETIC EVALUATION" .
If the value of the expression is non-zero, the return status is 0;
otherwise the return status is 1.  This is exactly equivalent to
\fBlet "\fIexpression\fP"\fR.
.TP
\fB[[\fP \fIexpression\fP \fB]]\fP
Return a status of 0 or 1 depending on the evaluation of
the conditional expression \fIexpression\fP.
Expressions are composed of the primaries described below under
.SM
.BR "CONDITIONAL EXPRESSIONS" .
Word splitting and pathname expansion are not performed on the words
between the \fB[[\fP and \fB]]\fP; tilde expansion, parameter and
variable expansion, arithmetic expansion, command substitution, process
substitution, and quote removal are performed.
Conditional operators such as \fB\-f\fP must be unquoted to be recognized
as primaries.
.if t .sp 0.5
.if n .sp 1
When the \fB==\fP and \fB!=\fP operators are used, the string to the
right of the operator is considered a pattern and matched according
to the rules described below under \fBPattern Matching\fP.
If the shell option
.B nocasematch
is enabled, the match is performed without regard to the case
of alphabetic characters.
The return value is 0 if the string matches (\fB==\fP) or does not match
(\fB!=\fP) the pattern, and 1 otherwise.
Any part of the pattern may be quoted to force it to be matched as a
string.
.if t .sp 0.5
.if n .sp 1
An additional binary operator, \fB=~\fP, is available, with the same
precedence as \fB==\fP and \fB!=\fP.
When it is used, the string to the right of the operator is considered
an extended regular expression and matched accordingly (as in \fIregex\fP(3)).  
The return value is 0 if the string matches
the pattern, and 1 otherwise.
If the regular expression is syntactically incorrect, the conditional
expression's return value is 2.
If the shell option
.B nocasematch
is enabled, the match is performed without regard to the case
of alphabetic characters.
Substrings matched by parenthesized subexpressions within the regular
expression are saved in the array variable \fBBASH_REMATCH\fP.
The element of \fBBASH_REMATCH\fP with index 0 is the portion of the string
matching the entire regular expression.
The element of \fBBASH_REMATCH\fP with index \fIn\fP is the portion of the
string matching the \fIn\fPth parenthesized subexpression.
.if t .sp 0.5
.if n .sp 1
Expressions may be combined using the following operators, listed
in decreasing order of precedence:
.if t .sp 0.5
.if n .sp 1
.RS
.PD 0
.TP
.B ( \fIexpression\fP )
Returns the value of \fIexpression\fP.

debian/changelog
vyatta-bash (3.2-4-vyatta-3) unstable; urgency=low

  * Remove AUDIT_SHELL option
  * Shell command logging support

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Fri, 18 Jun 2010 15:17:59 -0700

vyatta-bash (3.2-4-vyatta-2) unstable; urgency=low

  * Remove AUDIT_SHELL option

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Thu, 17 Jun 2010 10:17:47 -0700

vyatta-bash (3.2-4-vyatta-1) unstable; urgency=low

  * Don't remove vbash from /etc/shells during upgrade
  * Auditing is on for vbash
  * Fix command auditing to work right
  * Integrate bash 3.2 version

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Fri, 09 Apr 2010 16:17:17 -0700

vyatta-bash (3.1dfsg-8-vyatta-17) unstable; urgency=low

  * Add auditing support to bash
  * Fix spelling error warning
  * Turn on auditing for Vyatta bash
  * Package now depends on audit library
  * Rerun autoreconf to regenerate configure

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Wed, 07 Apr 2010 11:01:05 -0700

vyatta-bash (3.1dfsg-8-vyatta-16) unstable; urgency=low

  * Update FSF address
  * Add installation/removal scripts for vbash

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Wed, 10 Mar 2010 12:25:25 -0800

vyatta-bash (3.1dfsg-8-vyatta-15) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Wed, 17 Feb 2010 16:12:28 -0800

vyatta-bash (3.1dfsg-8-vyatta-14) unstable; urgency=low

  * Bugfix 4108: Check pointers before dereferencing them.

 -- Bob Gilligan <gilligan@vyatta.com>  Tue, 02 Jun 2009 14:14:02 -0700

vyatta-bash (3.1dfsg-8-vyatta-13) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Fri, 29 May 2009 18:34:45 -0700

vyatta-bash (3.1dfsg-8-vyatta-12) unstable; urgency=low

  Build system changes
  * add support for development build
  * update maintainer information
  * "files" file should be removed before package build

 -- An-Cheng Huang <ancheng@build-vm.localdomain>  Mon, 15 Dec 2008 16:07:17 -0800

vyatta-bash (3.1dfsg-8-vyatta-11) unstable; urgency=low

	  3.2.0
  [ Mark O'Brien ]


  [ slioch ]
  * remove ! and $ from vbash processing as these are handled by command
    scripts.
  * Handle double and single quote within a command argument. First
    quote turns on automatic escaping, second turns off. Will support
    unquoted, single quoted and
  * disable shell interpretation of ![num] when entering configure
    commands. this fixes a bug where the negation character in
    nat/firewall/wlb

  [ root ]
  * skipping whitespace at beginning of command when matching
    set|delete.
  * additional special characters to trap, added back free() after
    additional testing, removed generated files from project.
  * Revert "missing a free from last merge."
  * missing a free from last merge.
  * fix for bug 1283. Or partial fix. Now handle and escape special
    characters in vbash shell without requiring quotes on string. only
    applies to set/delete commands in vbash shell. Behavior is now much
    closer to juniper shell behavior with special characters.

  [ Mark O'Brien ]

 -- Mark O'Brien <mobrien@firebolt.vyatta.com>  Tue, 25 Nov 2008 19:08:33 -0800

vyatta-bash (3.1dfsg-8-vyatta-10) unstable; urgency=low

	  3.1.0
  [ Mark O'Brien ]


  [ rbalocca ]
  * Ignore derived files

  [ Mark O'Brien ]

 -- Mark O'Brien <mobrien@firebolt.vyatta.com>  Tue, 17 Jun 2008 09:25:53 -0700

vyatta-bash (3.1dfsg-8-vyatta-9) unstable; urgency=low

  3.0.5


 -- Mark O'Brien <mobrien@vyatta.com>  Tue, 06 May 2008 12:43:05 -0700

vyatta-bash (3.1dfsg-8-vyatta-8) unstable; urgency=low

  3.0.4


 -- Mark O'Brien <mobrien@vyatta.com>  Mon, 05 May 2008 16:40:24 -0700

vyatta-bash (3.1dfsg-8-vyatta-7) unstable; urgency=low

  3.0.3
  [ Mark O'Brien ]


  [ rbalocca ]
  * Indicate the VC4.0.2 release candidate in the changelog

  [ Mark O'Brien ]

 -- Mark O'Brien <mobrien@vyatta.com>  Tue, 29 Apr 2008 16:42:03 -0700

vyatta-bash (3.1dfsg-8-vyatta-6) unstable; urgency=low

  VC4.0.2


 -- Mark O'Brien <mobrien@vyatta.com>  Sat, 19 Apr 2008 11:55:52 -0700

vyatta-bash (3.1dfsg-8-vyatta-5) unstable; urgency=low

  VC4.0.2 release candidate


 -- Mark O'Brien <mobrien@vyatta.com>  Wed, 16 Apr 2008 09:49:46 -0700

vyatta-bash (3.1dfsg-8-vyatta-4) unstable; urgency=low

  3.0.2

  * 3.0.1

 -- Mark O'Brien <mobrien@vyatta.com>  Fri, 04 Apr 2008 18:00:09 -0700

vyatta-bash (3.1dfsg-8-vyatta-3) unstable; urgency=low

  VC4.0.1


 -- Mark O'Brien <mobrien@vyatta.com>  Tue, 18 Mar 2008 19:03:18 -0700

vyatta-bash (3.1dfsg-8-vyatta-2) unstable; urgency=low

  vc4.0.0
  [ Mark O'Brien ]


  [ An-Cheng Huang ]
  * fix for bug 2604: shell accepts '!' by default
  * fix for bug 2832: limit expansion checking to appropriate commands.

  [ rbalocca ]
  * Reset the version number to correspond to the version with which we
    started

  [ Mark O'Brien ]

 -- Mark O'Brien <mobrien@vyatta.com>  Mon, 25 Feb 2008 17:37:54 -0800

vyatta-bash (3.1dfsg-8-vyatta-1) unstable; urgency=low
  
  reset version #

 -- Rick Balocc <rbalocca@vyatta.com>  Fri, 14 Feb 2008 12:18:00 -0700

vyatta-bash (0.1) unstable; urgency=low
  
  * Initial release based on bash 3.1dfsg-8.

 -- An-Cheng Huang <ancheng@vyatta.com>  Mon, 12 Nov 2007 17:14:43 -0700

bash (3.1dfsg-8) unstable; urgency=high

  * bash-doc: Install copyright file when building without the
    gfdl'ed docs. Closes: #402471.

 -- Matthias Klose <doko@debian.org>  Mon, 11 Dec 2006 22:08:10 +0100

bash (3.1dfsg-7) unstable; urgency=medium

  * Restore the bash(1) manual page. Closes: #392989.

 -- Matthias Klose <doko@debian.org>  Sun, 22 Oct 2006 13:41:59 +0000

bash (3.1dfsg-6) unstable; urgency=low

  * Remove the texinfo documentation from the package. Closes: #357260.
  * Remove the FAQ from the package.

 -- Matthias Klose <doko@debian.org>  Thu, 12 Oct 2006 22:35:46 +0200

bash (3.1-5) unstable; urgency=low

  * Work around the broken remove-shell, not checking for the existance

debian/bash.preinst.c
/* Copyright (c) 1999 Anthony Towns
 * Copyright (c) 2000, 2002 Matthias Klose
 * Copyright (c) 2008 Canonical Ltd
 *
 * You may freely use, distribute, and modify this program.
 */

// Don't rely on /bin/sh and popen!

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/wait.h>

/* XXX: evil kludge, deal with arbitrary name lengths */
#ifndef PATH_MAX
#define PATH_MAX 4096
#endif

#include "md5.h"

int check_predepends(void)
{
    pid_t child;

    switch(child = fork()) {
      case -1:
        /* fork failed */
        return EXIT_FAILURE;

      case 0:
        /* i'm the child */
        {
            execl( "/usr/bin/dpkg", "/usr/bin/dpkg",
                   "--assert-support-predepends", NULL );
           _exit(127);
        }

      default:
        /* i'm the parent */
        {
            int status;
            pid_t pid;
            pid = wait(&status);
            if (pid == child) {
                if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
                    return EXIT_SUCCESS;
                }
            }
        }
    }
    return EXIT_FAILURE;
}

#define BUFSIZE 8192

int md5sum_match(char *fn, char* fn_digest)
{
  md5_state_t md5;
  md5_byte_t digest[16];
  unsigned char hexdigest[33];
  int i, j, fd;
  size_t nbytes;
  md5_byte_t buf[BUFSIZE];

  // if not existant, md5sums don't match
  if (access(fn, R_OK))
    return 0;
  if ((fd = open(fn, O_RDONLY)) == -1)
    return 0;
  
  md5_init(&md5);
  while (nbytes = read(fd, buf, BUFSIZE))
    md5_append(&md5, buf, nbytes);
  md5_finish(&md5, digest);
  close(fd);

  for (i = j = 0; i < 16; i++) {
    unsigned char c;
    c = (digest[i] >> 4) & 0xf;
    c = (c > 9) ? c + 'a' - 10 : c + '0';
    hexdigest[j++] = c;
    c = (digest[i] & 0xf);
    c = (c > 9) ? c + 'a' - 10 : c + '0';
    hexdigest[j++] = c;
  }
  hexdigest[j] = '\0';
#ifdef NDEBUG
  fprintf(stderr, "fn=%s, md5sum=%s, expected=%s\n", fn, hexdigest, fn_digest);
#endif
  return !strcmp(fn_digest, hexdigest);
}

int unmodified_file(char *fn, int md5sumc, unsigned char* md5sumv[])
{
  int i;

  // if not existant, pretend its unmodified
  if (access(fn, R_OK))
    return 1;
  for (i = 0; i < md5sumc; i++) {
    if (md5sum_match(fn, md5sumv[i]))
      return 1;
  }
  return 0;
}

unsigned char *md5sumv_bash_profile[] = {
  "d1a8c44e7dd1bed2f3e75d1343b6e4e1", // dapper, edgy, etch
  "0bc1802860b758732b862ef973cd79ff", // feisty, gutsy
};
const int md5sumc_bash_profile = sizeof(md5sumv_bash_profile) / sizeof (char *);

unsigned char *md5sumv_profile[] = {
  "7d97942254c076a2ea5ea72180266420", // feisty, gutsy
};
const int md5sumc_profile = sizeof(md5sumv_profile) / sizeof (char *);

#ifdef BC_CONFIG
unsigned char *md5sumv_completion[] = {
  "2bc0b6cf841eefd31d607e618f1ae29d", // dapper
  "ae1d298e51ea7f8253eea8b99333561f", // edgy
  "adc2e9fec28bd2d4a720e725294650f0", // feisty
  "c8bce25ea68fb0312579a421df99955c", // gutsy, and last one in bash
  "9da8d1c95748865d516764fb9af82af9", // etch, sid (last one in bash)
};
const int md5sumc_completion = sizeof(md5sumv_completion) / sizeof (char *);
#endif

char *check_diversion(void)
{
    pid_t child;
    int pipedes[2];

    if (pipe(pipedes))
	return NULL;


    switch(child = fork()) {
      case -1:
        /* fork failed */
	close(pipedes[0]);
	close(pipedes[1]);
        return NULL;

      case 0:
        /* i'm the child */
        {
	    if (dup2(pipedes[STDOUT_FILENO], STDOUT_FILENO) < 0)
		_exit(127);
	    close(pipedes[STDIN_FILENO]);
	    close(pipedes[STDOUT_FILENO]);
            execl( "/usr/sbin/dpkg-divert", "/usr/sbin/dpkg-divert",
                   "--list", "/bin/sh", NULL );
           _exit(127);
        }

      default:
        /* i'm the parent */
        {
	    static char line[1024];
	    FILE *fd;

	    close (pipedes[STDOUT_FILENO]);
	    fcntl (pipedes[STDIN_FILENO], F_SETFD, FD_CLOEXEC);
	    fd = fdopen (pipedes[STDIN_FILENO], "r");

	    while (fgets(line, 1024, fd) != NULL) {
		line[strlen(line)-1] = '\0';
		break;
	    }
	    fclose(fd);
	    return line;
        }
    }
    return NULL;
}

const char *msg =
  "As bash for Debian is destined to provide a working /bin/sh (pointing to\n"
  "/bin/bash) your link will be overwritten by a default link.\n\n"
  "If you don't want further upgrades to overwrite your customization, please\n"
  "read /usr/share/doc/bash/README.Debian.gz for a more permanent solution.\n\n"
  "[Press RETURN to continue]";

int main(void) {
    int targetlen;
    char target[PATH_MAX+1], answer[1024], *fn;

    fn = "/etc/skel/.bash_profile";
    if (!access(fn, R_OK)) {
      if (unmodified_file(fn, md5sumc_bash_profile, md5sumv_bash_profile)) {
	printf("removing %s in favour of /etc/skel/.profile\n", fn);
	unlink(fn);
      }
      else {
	fn = "/etc/skel/.profile";
	if (!access(fn, R_OK)) {
	  if (unmodified_file(fn, md5sumc_profile, md5sumv_profile)) {
	    printf("renaming /etc/skel/.bash_profile to %s\n", fn);
	    rename("/etc/skel/.bash_profile", fn);
	  }
	}
      }
    }

#ifdef BC_CONFIG
    fn = "/etc/bash_completion";
    if (!access(fn, R_OK)) {
      if (unmodified_file(fn, md5sumc_completion, md5sumv_completion)) {
	printf("removing unmodified %s, now in package bash-completion\n", fn);
	unlink(fn);
      }
    }
#endif

    if (check_predepends() != EXIT_SUCCESS) {
	printf("\nPlease upgrade to a new version of dpkg\n\n");
	return EXIT_FAILURE;
    }
    targetlen = readlink("/bin/sh", target, PATH_MAX);
    if (targetlen == -1) {
	// error reading link. Will be overwritten.
	puts("The bash upgrade discovered that something is wrong with your /bin/sh link.");
	puts(msg);
	fgets(answer, 1024, stdin);
	return EXIT_SUCCESS;
    }
    target[targetlen] = '\0';
    if (strcmp(target, "bash") != 0 && strcmp(target, "/bin/bash") != 0) {
	char *diversion = check_diversion();

	if (diversion == NULL)
	    return EXIT_FAILURE;
	// printf("diversion: /%s/\n", diversion);
	if (strcmp(diversion, "") != 0)
	    // link is diverted
	    return EXIT_SUCCESS;
	printf("The bash upgrade discovered that your /bin/sh link points to %s.\n", target);
	puts(msg);
	fgets(answer, 1024, stdin);
	return EXIT_SUCCESS;
    }

    return EXIT_SUCCESS;
}
